client/thin-replica-client/src/grpc_connection.cpp

```diff
@@ -356,26 +374,131 @@ TrsConnection::Result TrsConnection::readHash(Hash* hash) {
+    snapshot_context->TryCancel();
+    stream.wait();
+    snapshot_context.reset();
+
+    // If StreamSnapshot did end up returning a pointer to an allocated
+    // stream, make sure it does not get leaked.
+    snapshot_stream = stream.get();
+    snapshot_stream.reset();
+    current_req_id_.fetch_sub(1u);
+    request_id = 0u;
+    return Result::kTimeout;
+  }
+
+  ConcordAssert(status == future_status::ready);
+  snapshot_stream = stream.get();
+  if (snapshot_stream) {
+    std::lock_guard lock(rss_streams_mutex_);
+    rss_streams_.emplace(request_id, std::make_pair(std::move(snapshot_context), std::move(snapshot_stream)));
+    return Result::kSuccess;
+  }
+
+  current_req_id_.fetch_sub(1u);
+  request_id = 0u;
+  snapshot_stream.reset();
+  snapshot_context.reset();
   return Result::kFailure;
 }
 
-void TrsConnection::cancelStateSnapshotStream() {
-  // TODO: Add implementation
-  ConcordAssert("cancelStateSnapshotStream should not be called. It is unimplemented." && false);
-  return;
+void GrpcConnection::cancelStateSnapshotStream(RequestId request_id) {
+  std::lock_guard lock(rss_streams_mutex_);
+  auto it = rss_streams_.find(request_id);
+  if (it != rss_streams_.end()) {
+    if (!(it->second).second) {
+      ConcordAssertEQ((it->second).first, nullptr);
+      return;
+    }
+    ConcordAssertNE((it->second).first, nullptr);
+    (it->second).first->TryCancel();
+    (it->second).first.reset();
+    (it->second).second.reset();
+  }
+  rss_streams_.erase(request_id);
+}
+
+void GrpcConnection::cancelStateSnapshotStream() {
+  std::lock_guard lock(rss_streams_mutex_);
+  for (auto& c : rss_streams_) {
+    if (!c.second.second) {
+      ConcordAssertEQ(c.second.first, nullptr);
+      continue;
+    }
+    ConcordAssertNE(c.second.first, nullptr);
+    c.second.first->TryCancel();
+    c.second.first.reset();
+    c.second.second.reset();
+  }
+  rss_streams_.clear();
 }
-bool TrsConnection::hasStateSnapshotStream() {
-  // TODO: Add implementation
-  ConcordAssert("hasStateSnapshotStream should not be called. It is unimplemented." && false);
+
+bool GrpcConnection::hasStateSnapshotStream(RequestId request_id) {
+  std::lock_guard lock(rss_streams_mutex_);
+  auto it = rss_streams_.find(request_id);
+  if (it != rss_streams_.end()) {
+    return bool((it->second).second);
+  }
   return false;
 }
-TrsConnection::Result TrsConnection::readStateSnapshot(vmware::concord::replicastatesnapshot::KeyValuePair* key_value) {
-  // TODO: Add implementation
-  ConcordAssert("readStateSnapshot should not be called. It is unimplemented." && false);
+
+GrpcConnection::Result GrpcConnection::readStateSnapshot(
+    RequestId request_id, vmware::concord::replicastatesnapshot::StreamSnapshotResponse* snapshot_response) {
+  auto result = async(launch::async, [this, snapshot_response, request_id] {
+    std::lock_guard lock(this->rss_streams_mutex_);
+    return (((this->rss_streams_)[request_id]).second)->Read(snapshot_response);
```

Comment:
We are holding the lock on **ALL** stream objects for the `Read` call on a single stream object. If the call times out then we were holding the lock for default 5 seconds. Obviously, we need to prevent anyone from cleaning up the stream object but isn't there a better way? Maybe it won't be a practical issue but holding a lock for 5s on all streams doesn't seem Ok to me.

Reply:
This will depend on the load. If we will see that there are many clients accessing and there are many timeouts then this will be a common scenario. But as for most of the call Read is a read of one datum, it will not be 5s and there should not be an issue in practice.